src/hotspot/share/opto/output.cpp

```diff
@@ -1683,6 +1694,8 @@ void PhaseOutput::fill_buffer(CodeBuffer* cb, uint* blk_starts) {
       n->emit(*cb, C->regalloc());
       current_offset = cb->insts_size();
 
+      assert(!is_mcall || (call_returns[block->_pre_order] == (uint) current_offset), "ret_addr_offset() did not match size of emitted code");
```

Comment:
This assertion is too strong: on AArch64 we generate additional instructions after the BLR (call) instruction for certain types of call. For example

```  
  0x0000ffff790f00dc:   adr     x9, 0x0000ffff790f00f4
  0x0000ffff790f00e0:   mov     x8, #0x5714                     // #22292
  0x0000ffff790f00e4:   movk    x8, #0x8d3d, lsl #16
  0x0000ffff790f00e8:   movk    x8, #0xffff, lsl #32
  0x0000ffff790f00ec:   stp     xzr, x9, [sp, #-16]!
  0x0000ffff790f00f0:   blr     x8
  0x0000ffff790f00f4:   add     sp, sp, #0x10                          <== ret_addr_offset() is here
  0x0000ffff790f00f8:   Address 0x0000ffff790f00f8 is out of bounds.   <== current_offset is here
```

I think the `==` should be `<=`. (Although this still fails sometimes on AArch64, but I believe it exposes a real bug. I've opened JDK-8256025 and will fix this shortly.)

Reply:
Ok, that seems fine to me. IIRC the problem this was trying to catch is a ret_addr_offset that is too large, which might cause a later call's oop map to be overridden. So, using `<=` should still work. At least if the code between ret_addr_offset and current_offset is guaranteed not to contain any calls or other safepoints.